### PR TITLE
Add fltlib API

### PR DESCRIPTION
**Pull request by mat-gas**
this PR adds support for fltlib API :
  - https://docs.microsoft.com/en-us/windows/win32/api/fltuser/ 
  - https://docs.microsoft.com/en-us/windows-hardware/drivers/ifs/filter-manager-concepts

This way, winapi is (normally) able to load/unload, connect to a minifilter..

### CODE REVIEW

File: src/um/fltuser.rs [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        lpInBuffer: LPVOID,
+        dwInBufferSize: DWORD,
+        lpOutBuffer: LPVOID,
+        dwOutBufferSize: DWORD,
+        lpBytesReturned: LPDWORD
+    ) -> HRESULT;
+    pub fn FilterUnload(
+        lpFilterName: LPCWSTR
+    ) -> HRESULT;
+    // FilterVolumeClose@4 // ?????
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
```

Not sure what this comment is about? I couldn't find FilterVolumeClose in the header.

File: src/um/fltuser.rs [hunk 2]
```diff
@@ -0,0 +1,376 @@
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// All files in the project carrying such notice may not be copied, modified, or distributed
+// except according to those terms.
+use shared::minwindef::{DWORD, LPCVOID, LPDWORD, LPHANDLE, LPVOID, ULONG, USHORT, WORD};
+use shared::ntdef::NTSTATUS;
+use um::minwinbase::{LPOVERLAPPED, LPSECURITY_ATTRIBUTES};
+use um::winnt::{HANDLE, HRESULT, LPCWSTR, LPWSTR, PHANDLE, ULONGLONG, WCHAR};
+pub type HFILTER = HANDLE;
```
Comment (roblabla) [inline src/um/fltuser.rs]: A lot of those definitions should go in a module `shared::fltuserstructures`, as they are defined in a header called `shared/fltuserstructures.h`

File: src/um/fltuser.rs [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+pub type PHFILTER = *mut HFILTER;
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
```

I couldn't find `PHFILTER` in the windows header files.

File: src/um/fltuser.rs [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+pub type HFILTER_INSTANCE = HANDLE;
+pub type PHFILTER_INSTANCE = *mut HFILTER_INSTANCE;
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
```

I couldn't find `PHFILTER_INSTANCE` in the windows header files.

File: src/um/fltuser.rs [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    FLT_FSTYPE_GPFS,            //IBM General Parallel File System  (\FileSystem\gpfs)
+    FLT_FSTYPE_NPFS,            //Microsoft's Named Pipe file system(\FileSystem\npfs)
+    FLT_FSTYPE_MSFS,            //Microsoft's Mailslot file system  (\FileSystem\msfs)
+    FLT_FSTYPE_CSVFS,           //Microsoft's Cluster Shared Volume file system  (\FileSystem\csvfs)
+    FLT_FSTYPE_REFS,            //Microsoft's ReFS file system      (\FileSystem\Refs or \FileSystem\Refsv1)
+    FLT_FSTYPE_OPENAFS,         //OpenAFS file system               (\Device\AFSRedirector)
+    FLT_FSTYPE_CIMFS,            //Composite Image file system       (\FileSystem\cimfs)
+}}
+ENUM!{enum FILTER_INFORMATION_CLASS {
+    FilterFullInformation = 0,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    FilterFullInformation,
```

Windows header file doesn't specify `= 0`.

File: src/um/fltuser.rs [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+    FilterAggregateBasicInformation,
+    FilterAggregateStandardInformation,
+}}
+ENUM!{enum FILTER_VOLUME_INFORMATION_CLASS {
+    FilterVolumeBasicInformation = 0,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    FilterVolumeBasicInformation,
```

Windows header file doesn't specify = 0.

File: src/um/fltuser.rs [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+    FilterVolumeStandardInformation,
+}}
+ENUM!{enum INSTANCE_INFORMATION_CLASS {
+    InstanceBasicInformation = 0,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    InstanceBasicInformation,
```

Windows header file doesn't specify `= 0`.

File: src/um/fltuser.rs [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    FilterAltitudeLength: USHORT,
+    FilterAltitudeBufferOffset: USHORT,
+}}
+STRUCT!{struct FILTER_AGGREGATE_BASIC_INFORMATION_u_s_LegacyFilter {
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+}}
+UNION!{union FILTER_AGGREGATE_BASIC_INFORMATION_Type_u {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
UNION!{union FILTER_AGGREGATE_BASIC_INFORMATION_Type {
```

From the contributor guide, on handling anonymous structs/unions:

> If the field does have a name, such as Bar, but still contains an anonymous struct or union then retain the name of Bar for the field and name the anonymous struct or union after the field, such as FOO_Bar.

File: src/um/fltuser.rs [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    InstanceAggregateStandardInformation,
+}}
+STRUCT!{struct FILTER_FULL_INFORMATION {
+    NextEntryOffset: ULONG,
+    FrameID: ULONG,
+    NumberOfInstances: ULONG,
+    FilterNameLength: USHORT,
+    FilterNameBuffer: [WCHAR; 1],
+}}
+STRUCT!{struct FILTER_AGGREGATE_BASIC_INFORMATION_u_s_MiniFilter {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct FILTER_AGGREGATE_BASIC_INFORMATION_Type_MiniFilter {
```

From the contributor guide, on handling anonymous structs/unions:

File: src/um/fltuser.rs [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+    FrameID: ULONG,
+    NumberOfInstances: ULONG,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    FilterAltitudeLength: USHORT,
+    FilterAltitudeBufferOffset: USHORT,
+}}
+STRUCT!{struct FILTER_AGGREGATE_BASIC_INFORMATION_u_s_LegacyFilter {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct FILTER_AGGREGATE_BASIC_INFORMATION_Type_LegacyFilter {
```

From the contributor guide, on handling anonymous structs/unions:

> If the field does have a name, such as Bar, but still contains an anonymous struct or union then retain the name of Bar for the field and name the anonymous struct or union after the field, such as FOO_Bar.

File: src/um/fltuser.rs [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+}}
+UNION!{union FILTER_AGGREGATE_BASIC_INFORMATION_Type_u {
+    [u32; 4],
+    MiniFilter mut_MiniFilter: FILTER_AGGREGATE_BASIC_INFORMATION_u_s_MiniFilter,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    MiniFilter MiniFilter_mut: FILTER_AGGREGATE_BASIC_INFORMATION_Type_MiniFilter,
```

The name of the mutable accessor should be `FieldName_mut` with a `_mut` suffix.

File: src/um/fltuser.rs [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+    LegacyFilter mut_LegacyFilter: FILTER_AGGREGATE_BASIC_INFORMATION_u_s_LegacyFilter,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    LegacyFilter LegacyFilter_mut: FILTER_AGGREGATE_BASIC_INFORMATION_Type_LegacyFilter,
```

File: src/um/fltuser.rs [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+}}
+STRUCT!{struct FILTER_AGGREGATE_BASIC_INFORMATION {
+    NextEntryOffset: ULONG,
+    Flags: ULONG,
+    Type: FILTER_AGGREGATE_BASIC_INFORMATION_Type_u,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    Type: FILTER_AGGREGATE_BASIC_INFORMATION_Type,
```

File: src/um/fltuser.rs [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+}}
+STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_MiniFilter {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION_Type_MiniFilter {
```

File: src/um/fltuser.rs [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+    Flags: ULONG,
+    FrameID: ULONG,
+    NumberOfInstances: ULONG,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    FilterAltitudeLength: USHORT,
+    FilterAltitudeBufferOffset: USHORT,
+}}
+STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_LegacyFilter {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION_Type_LegacyFilter {
```

File: src/um/fltuser.rs [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    Flags: ULONG,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    FilterAltitudeLength: USHORT,
+    FilterAltitudeBufferOffset: USHORT,
+}}
+UNION!{union FILTER_AGGREGATE_STANDARD_INFORMATION_Type_u {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
UNION!{union FILTER_AGGREGATE_STANDARD_INFORMATION_Type {
```

File: src/um/fltuser.rs [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+}}
+STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION {
+    NextEntryOffset: ULONG,
+    Flags: ULONG,
+    Type: FILTER_AGGREGATE_STANDARD_INFORMATION_Type_u,
+}}
+STRUCT!{struct FILTER_VOLUME_BASIC_INFORMATION {
+    FilterVolumeNameLength: USHORT,
+    FilterVolumeName: [WCHAR; 1],
+ }}
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
}}
```

File: src/um/fltuser.rs [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+ STRUCT!{struct FILTER_VOLUME_STANDARD_INFORMATION {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct FILTER_VOLUME_STANDARD_INFORMATION {
```

File: src/um/fltuser.rs [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+    NextEntryOffset: ULONG,
+    Flags: ULONG,
+    FrameID: ULONG,
+    FileSystemType: FLT_FILESYSTEM_TYPE,
+    FilterVolumeNameLength: USHORT,
+    FilterVolumeName: [WCHAR; 1],
+ }}
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
}}
```

File: src/um/fltuser.rs [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+}}
+UNION!{union FILTER_AGGREGATE_STANDARD_INFORMATION_Type_u {
+    [u32; 5],
+    MiniFilter mut_MiniFilter: FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_MiniFilter,
+    LegacyFilter mut_LegacyFilter: FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_LegacyFilter,
+}}
+STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION {
+    NextEntryOffset: ULONG,
+    Flags: ULONG,
+    Type: FILTER_AGGREGATE_STANDARD_INFORMATION_Type_u,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    Type: FILTER_AGGREGATE_STANDARD_INFORMATION_Type,
```

File: src/um/fltuser.rs [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+STRUCT!{struct FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_LegacyFilter {
+    Flags: ULONG,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    FilterAltitudeLength: USHORT,
+    FilterAltitudeBufferOffset: USHORT,
+}}
+UNION!{union FILTER_AGGREGATE_STANDARD_INFORMATION_Type_u {
+    [u32; 5],
+    MiniFilter mut_MiniFilter: FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_MiniFilter,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    MiniFilter MiniFilter_mut: FILTER_AGGREGATE_STANDARD_INFORMATION_Type_MiniFilter,
```

File: src/um/fltuser.rs [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+    LegacyFilter mut_LegacyFilter: FILTER_AGGREGATE_STANDARD_INFORMATION_u_s_LegacyFilter,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    LegacyFilter LegacyFilter_mut: FILTER_AGGREGATE_STANDARD_INFORMATION_Type_LegacyFilter,
```

File: src/um/fltuser.rs [hunk 23]
```diff
@@ -0,0 +1,376 @@
...
+    InstanceNameBufferOffset: USHORT,
+    AltitudeLength: USHORT,
+    AltitudeBufferOffset: USHORT,
+    VolumeNameLength: USHORT,
+    VolumeNameBufferOffset: USHORT,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    SupportedFeatures: ULONG,
+}}
+STRUCT!{struct INSTANCE_AGGREGATE_STANDARD_INFORMATION_u_s_LegacyFilter {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_LegacyFilter {
```

File: src/um/fltuser.rs [hunk 24]
```diff
@@ -0,0 +1,376 @@
...
+    Flags: ULONG,
+    AltitudeLength: USHORT,
+    AltitudeBufferOffset: USHORT,
+    VolumeNameLength: USHORT,
+    VolumeNameBufferOffset: USHORT,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    SupportedFeatures: ULONG,
+}}
+UNION!{union INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_u {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
UNION!{union INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type {
```

File: src/um/fltuser.rs [hunk 25]
```diff
@@ -0,0 +1,376 @@
...
+    InstanceNameLength: USHORT,
+    InstanceNameBufferOffset: USHORT,
+    AltitudeLength: USHORT,
+    AltitudeBufferOffset: USHORT,
+    VolumeNameLength: USHORT,
+    VolumeNameBufferOffset: USHORT,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+}}
+STRUCT!{struct INSTANCE_AGGREGATE_STANDARD_INFORMATION_u_s_MiniFilter {
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
STRUCT!{struct INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_MiniFilter {
```

File: src/um/fltuser.rs [hunk 26]
```diff
@@ -0,0 +1,376 @@
...
+    AltitudeBufferOffset: USHORT,
+    VolumeNameLength: USHORT,
+    VolumeNameBufferOffset: USHORT,
+    FilterNameLength: USHORT,
+    FilterNameBufferOffset: USHORT,
+    SupportedFeatures: ULONG,
+}}
+UNION!{union INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_u {
+    [u32; 8],
+    MiniFilter mut_MiniFilter: INSTANCE_AGGREGATE_STANDARD_INFORMATION_u_s_MiniFilter,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    MiniFilter MiniFilter_mut: INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_MiniFilter,
```

File: src/um/fltuser.rs [hunk 27]
```diff
@@ -0,0 +1,376 @@
...
+    LegacyFilter mut_LegacyFilter: INSTANCE_AGGREGATE_STANDARD_INFORMATION_u_s_LegacyFilter,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    LegacyFilter LegacyFilter_mut: INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_LegacyFilter,
```

File: src/um/fltuser.rs [hunk 28]
```diff
@@ -0,0 +1,376 @@
...
+}}
+STRUCT!{struct INSTANCE_AGGREGATE_STANDARD_INFORMATION {
+    NextEntryOffset: ULONG,
+    Flags: ULONG,
+    Type: INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type_u,
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
    Type: INSTANCE_AGGREGATE_STANDARD_INFORMATION_Type,
```

File: src/um/fltuser.rs [hunk 29]
```diff
@@ -0,0 +1,376 @@
...
+    pub fn FilterClose(
+        hFilter: HFILTER
+    ) -> HRESULT;
+    pub fn FilterConnectCommunicationPort(
+        lpPortName: LPCWSTR,
+        dwOptions: DWORD,
+        lpContext: LPCVOID,
+        wSizeOfContext: WORD,
+        lpSecurityAttributes: LPSECURITY_ATTRIBUTES,
+        hPort: PHANDLE
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
        hPort: *mut HANDLE
```

The windows header file uses a raw pointer here.

File: src/um/fltuser.rs [hunk 30]
```diff
@@ -0,0 +1,376 @@
...
+    ) -> HRESULT;
+    pub fn FilterCreate(
+        lpFilterName: LPCWSTR,
+        hFilter: PHFILTER
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
        hFilter: *mut HFILTER
```

File: src/um/fltuser.rs [hunk 31]
```diff
@@ -0,0 +1,376 @@
...
+        lpOverlapped: LPOVERLAPPED
+    ) -> HRESULT;
+    pub fn FilterInstanceClose(
+        hInstance: HFILTER_INSTANCE
+    ) -> HRESULT;
+    pub fn FilterInstanceCreate(
+        lpFilterName: LPCWSTR,
+        lpVolumeName: LPCWSTR,
+        lpInstanceName: LPCWSTR,
+        hInstance: PHFILTER_INSTANCE
```
Comment (roblabla) [inline src/um/fltuser.rs]:
```suggestion
        hInstance: *mut HFILTER_INSTANCE
```

**Review by roblabla — CHANGES_REQUESTED** [review]
Requires changes.

- Organization of code
  - All definitions go into the source file that directly maps to the header the definition is from:
    - :warning: some definitions in fltuser should go in a separate `shared::fltuserstructures` module.
  - Definitions are defined in the same order as they are in the original header
    - :warning: Definitions are not given in the same order as the original header.

- Extern functions
  - ABIS are correct (WINAPI = "system"): :heavy_check_mark: 
  - Names are correct (functions, parameters): :heavy_check_mark: 
  - Parameter types are correct: :warning: PHANDLE used instead of *mut HANDLE (resolves to the same underlying type).
  - Return types are correct: :heavy_check_mark: 

- Constants
  - Names are correct: :heavy_check_mark: 
  - Values are correct: :heavy_check_mark: 
  - Types are correct: :heavy_check_mark: 
    - All the ULONG consts go in various ULONG members of structures.
    - `FLT_PORT_FLAG_SYNC_HANDLE` is used with `FilterConnectCommunicationPort`' dwOptions, a DWORD.

- Structs
  - Struct definitions are wrapped in `STRUCT!` macro: :heavy_check_mark: 
  - Struct names are correct: :warning: problems around anonymous names
  - Param names are correct: :heavy_check_mark: 
  - Param types are correct: :heavy_check_mark: 

- Unions
  - Unions are wrapped in the `UNION!` macro: :heavy_check_mark: 
  - Union names are correct: :warning: problems around anonymous unions 
  - Union container size is correct: :heavy_check_mark: 
  - Union member names are correct: :heavy_check_mark: 
  - Union member mutable accessor is in form `FieldName_mut`: :warning: 
  - Union member types are correct: :heavy_check_mark: 

- Enums
  - Enum definitions are wrapped in `ENUM!` macro: :heavy_check_mark: 
  - Enum names are correct: :heavy_check_mark: 
  - Enum variant names are correct: :heavy_check_mark: 
  - Enum variant values are correct: :heavy_check_mark:

**Review by roblabla — APPROVED** [review]
LGTM :+1:

All problems from previous review fixed!

**Comment by mvforell** [comment]
Any estimate on when this will be merged and when a new version will be published?

**Pull request closed without merging**